module/zfs/zvol.c

```diff
@@ -966,13 +968,22 @@ zvol_request(struct request_queue *q, struct bio *bio)
 		 */
 		zvr->rl = zfs_range_lock(&zv->zv_range_lock, offset, size,
 		    RL_WRITER);
+		/*
+		 * FIXME: if there is at least another lock held on the same
+		 * block process this request synchronously.
+		 */
+		if (zfs_range_locked(&zv->zv_range_lock, P2ALIGN(offset,
+		    zv->zv_volblocksize), zv->zv_volblocksize) > 1)
+			need_sync = B_TRUE;
```

Comment:
This works but I worry about the impact on performance.  You should be able to resolve this circular dependency similarly by making the sync check outside `zvol_write()`.  This way the `zil_commit()` never gets run in the taskq thread where it might block consuming the thread.  Incidentally, this is why the `zfs_range_lock()` is always acquired in `zvol_request()`.

```
need_sync = bio_is_fua(bio) || zv->zv_objset->os_sync == ZFS_SYNC_ALWAYS;
```